sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::time::Duration;
+use thiserror::Error;
+
+/// Error related to the `Emulator` trait. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "Emulator error: {}".  So including the words "error" or
+/// "Emulator" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Timeout during performing command: {0} timeout: {1}")]
+    Timeout(String, u64),
+    #[error("Can't restore resource to initial state: {0}")]
+    ResetError(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum EmuValue {
+    Empty,
+    String(String),
+    FilePath(String),
+    StringList(Vec<String>),
+    FilePathList(Vec<String>),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a `Emulator` instance
+pub trait Emulator {
+    /// Acquires state of emulator
+    fn get_state(&self) -> Result<EmuState>;
+
+    /// Waits for the specified amount of time passed in `timeout`, for first state transition
+    /// of the Emulator form `Busy` to one of states defined in `EmuState` other that `Busy`.
+    /// If the transition from `Busy` to another state does not take place until
+    /// the `timeout` expires, the function will return an `Timeout` error.
+    /// In case of any transition of Emulator to `Error` state,
+    /// the function immediately returns `RuntimeError` with description of a problem.
+    fn wait_until_busy(&self, timeout: Duration) -> Result<()>;
+
+    /// Start Emulator with provided arguments. Parameter `timeout`
+    /// describe how long Emulator will wait for emulation to start.
+    /// The `reset_resource` parameter describes whether the resources are to be restored
+    /// to their initial state before starting emulated DUT.
+    /// If the parameter value is true, the resource file will be restored to its original content.
```

Comment:
It is unclear to me what is meant by "its original content".  For instance, will the EEPROM be set to all zeroes, or will it be set to the same content that was given in an args entry at a previous invocation of start()?
Also, I assume that reset_resource only has an effect on resources that are not explicitly mentioned in the args map.

Stepping back, just to make sure we are on the same page.  If I call start() twice, with the same filename in the EEPROM key twice, the file will not have been modified by the emulation, and the effect is that the second invocation of start() will reset the content of the EEPROM to be identical to what it was after the first invocation of start().  Right?

Reply:
In this case original content mean - first send content unmodified by invocation of start. 
**EEPROM to be identical to what it was after the first invocation of start(). Right?** Yes

Reply:
About other things not mentioned in `args` - it should be implementation dependent.

Reply:
Michał and I discussed this, and agreed on the following convention:
If a resource (eeprom, fuses, flash, etc.) is mentioned in the `args` set, it will be "updated" to match the content of that file, that is, any modifications to e.g. eeprom that may have taken place during previous emulation run are discarded. (And the emulator will never modify files provided to `start()`, but if in need to persist modifications by the emulated code on disk, create separate temporary files or employ "copy-on-write" or similar.)
If not mentioned, and `reset` is false, then the content including modifications by previous emulation runs are kept.
If not mentioned, and `reset` is true, then the content is reset to all zeroes, (or whatever was the initial state for that particular resource before the very first invocation of `start()`.)